### PR TITLE
(PDB-5095) Time shift PuppetDB export

### PR DESCRIPTION
**Pull request by Filipovici-Andrei**
The rest of the implementation for the time shift of PuppetDB export archive, utility.

### CODE REVIEW

**Review by rbrw — COMMENTED** [review]
I didn't quite finish the review pass on Friday, but here's what I had, and I should have time to finish on Monday.

File: src/puppetlabs/puppetdb/cli/time_shift_export.clj [hunk 1]
```diff
@@ -4,59 +4,233 @@
...
+   [clojure.java.io :as io]
+   [puppetlabs.puppetdb.archive :as archive]
+   [puppetlabs.puppetdb.time :refer [parse-wire-datetime after? now interval plus]]
+   [puppetlabs.puppetdb.cheshire :as json]
+   [clojure.string :as str]))
+
+(defn time-shift-timestamp
+  [timestamp-string shift-period]
+
+  (-> timestamp-string
```
[the remark this answers is not in the document]
Comment (Filipovici-Andrei) [reply src/puppetlabs/puppetdb/cli/time_shift_export.clj:31]: Removed the empty line

File: src/puppetlabs/puppetdb/cli/time_shift_export.clj [hunk 2]
```diff
@@ -4,59 +4,233 @@
...
+      (plus shift-period)))
+
+(defn time-shift-resources
+  [resources shift-period]
+
+  (mapv
+    (fn [resource]
+      (-> resource
+          (update :timestamp #(time-shift-timestamp % shift-period))
+          (update :events (fn[events]
```
[the remark this answers is not in the document]
Comment (Filipovici-Andrei) [reply src/puppetlabs/puppetdb/cli/time_shift_export.clj:42]: I'm not sure what you mean.

File: test/puppetlabs/puppetdb/integration/time_shift_export.clj [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+          metadata "export-metadata.json"
+          input-archive-path "dev-resources/timeshift-export/input-archive"]
+
+      (println (shell/sh "tar" "-czvf" input-archive (str "--directory=" input-archive-path) catalog fact report metadata))
+      (is (.exists (io/file input-archive)))
+
+      (time-shift "-i" input-archive "-o" output-archive "-t" "2021-05-14T22:22:00.000Z")
+      (is (.exists (io/file output-archive)))
+
+      (shell/sh "tar" "-xvzf" output-archive)
```
[the remark this answers is not in the document]
Comment (Filipovici-Andrei) [reply test/puppetlabs/puppetdb/integration/time_shift_export.clj:44]: Should I apply this just for the `tar` commands?

File: src/puppetlabs/puppetdb/cli/time_shift_export.clj [hunk 4]
```diff
@@ -4,59 +4,233 @@
...
-   input export but with the 'shifted' prefix."
-
+   It accepts three parameters:
+    - [Mandatory] -i / --input
+      Path to the .tgz pdb export, which will be shifted.
+    - [Optional] -o / --output
+      Path to the where the shifted export will be saved. If no path is given,
+      the archive will be stored in the same place as the input export,
+      but the export name will be prefixed with \"shifted\". (e.g input-export.tgz will become shifted-input-export.tgz).
+       If the output path doesn't end with .tgz it will be added.
```
[the remark this answers is not in the document]
Comment (Filipovici-Andrei) [reply src/puppetlabs/puppetdb/cli/time_shift_export.clj:14]: Yes, we tarball-writer does allow writing to standard output. Modified that part, as suggested.

File: src/puppetlabs/puppetdb/cli/time_shift_export.clj [hunk 5]
```diff
@@ -4,32 +4,174 @@
...
+       (reduce get-newest-timestamp)))
+
+(defn newest-timestamp-from-report
+  [json-data]
+  (->> (concat [(parse-wire-datetime (json-data :producer_timestamp))
+                (parse-wire-datetime (json-data :start_time))
+                (parse-wire-datetime (json-data :end_time))]
+               (map #(parse-wire-datetime (:time %)) (json-data :logs)))
+       (reduce get-newest-timestamp)
+       (get-newest-timestamp (resources-newest-timestamp (json-data :resources)))))
```
Comment (Filipovici-Andrei) [inline src/puppetlabs/puppetdb/cli/time_shift_export.clj:111]: The current version might be faster, but I think this 
`(defn get-resources-timestamps
  [resources]
  (mapcat (fn [resource]
            (concat [(resource :timestamp)]
                    (map #(:timestamp %) (resource :events))))
          resources))

(defn newest-timestamp-from-report
  [json-data]
  (->> (concat [(json-data :producer_timestamp)
                (json-data :start_time)
                (json-data :end_time)]
               (map #(:time %) (json-data :logs))
               (get-resources-timestamps (json-data :resources)))
       (map parse-wire-datetime)
       (reduce get-newest-timestamp)))` 
looks better

**Pull request merged**